src/main/java/software/amazon/smithy/lsp/handler/DefinitionHandler.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.lsp.handler;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import org.eclipse.lsp4j.DefinitionParams;
+import org.eclipse.lsp4j.Location;
+import org.eclipse.lsp4j.Position;
+import software.amazon.smithy.lsp.document.DocumentId;
+import software.amazon.smithy.lsp.document.DocumentParser;
+import software.amazon.smithy.lsp.document.DocumentPositionContext;
+import software.amazon.smithy.lsp.project.Project;
+import software.amazon.smithy.lsp.project.SmithyFile;
+import software.amazon.smithy.lsp.protocol.LocationAdapter;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.loader.Prelude;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.MixinTrait;
+import software.amazon.smithy.model.traits.TraitDefinition;
+import software.amazon.smithy.model.validation.ValidatedResult;
+
+public final class DefinitionHandler {
+    private final Project project;
+
+    public DefinitionHandler(Project project) {
+        this.project = project;
+    }
+
+    /**
+     * @param params The request params
+     * @return A list of possible definition locations
+     */
+    public List<Location> handle(DefinitionParams params) {
+        String uri = params.getTextDocument().getUri();
+        SmithyFile smithyFile = project.getSmithyFile(uri);
+        if (smithyFile == null) {
```

Comment:
Hmm yea this should probably be a null check on the `project`, since we go from uri in the request -> project; this might be an artifact from when I initially did the refactor and didn't handle 'detached' files. 

But generally speaking, there's probably some room for improvement here. For these kinds of text document features, the server won't be receiving requests for files that haven't been opened in the client, and any time a document is opened in the client, we receive `didOpen` (LSP does ensure request ordering), and we always create a corresponding project. Because of that, maybe the right thing to do is actually to throw if we receive a request for a file that the server knows nothing about.

Reply:
>  maybe the right thing to do is actually to throw if we receive a request for a file that the server knows nothing about.

This seems right to me

Reply:
TODO should at least log something